Refactor code in groups/ to make it mockable

MadhavJivrajani opened the pull request:
- Introduce mockable client interfaces that represent interaction with the admins API.
- Introduce mockable service interfaces that represent high level operations that are
  performed during reconcilation.
- Misc: fix a few warnings and linter errors

Part of #420

MadhavJivrajani commented:
/cc @nikhita
fyi

MadhavJivrajani commented:
/test all

spiffxp commented on groups/reconcile.go:
```diff
@@ -181,69 +184,176 @@ func main() {
+	return nil
+}
+
+func (r *reconciler) printGroupMembersAndSettings() error {
```
What would make this easier to review is to leave this function where it was instead of reordering

Make whatever changes you're going to make to the function

Then reorder in a new commit

spiffxp commented on groups/reconcile.go:
```diff
@@ -311,12 +421,19 @@ func (gc *GroupsConfig) Load(rootDir string, restrictions *RestrictionsConfig) e
 	log.Printf("reading groups.yaml files recursively at %s", rootDir)
 
 	return filepath.Walk(rootDir, func(path string, info os.FileInfo, err error) error {
+		// check if the current path is accessible or not.
+		if err != nil {
+			return err
+		}
```
This looks like new logic.  Ideally we'd see this in a separate commit from a refactor, with a test case that proves why it's needed

spiffxp commented on groups/reconcile.go:
```diff
@@ -311,12 +421,19 @@ func (gc *GroupsConfig) Load(rootDir string, restrictions *RestrictionsConfig) e
 
-			var groupsConfigAtPath GroupsConfig
-			var content []byte
+			var (
+				groupsConfigAtPath GroupsConfig
+				content            []byte
+				err                error
+			)
```
I am not used to seeing this style of var declaration outside of globals but maybe that's just me

spiffxp commented on groups/reconcile.go:
```diff
@@ -181,69 +184,176 @@ func main() {
 		}
 
-		err = createOrUpdateGroupIfNecessary(adminService, g.EmailId, g.Name, g.Description)
+		// update the group that is currently being considered.
+		r.adminService.SetGroup(g)
+		r.groupService.SetGroup(g)
```
This context-aware API seems prone to human error.  I would rather either:
- make group an explicit parameter to function calls (e.g. `adminService.CreateOrUpdateGroupIfNescessary(g)`)
- return and use an instance of something that is configured to use group (e.g. `a = adminServiceFactory.AdminServiceForGroup(g)` but those are terrible names I'm sure you can think of something better)

spiffxp commented on groups/service.go:
```diff
@@ -0,0 +1,415 @@
+	"google.golang.org/api/option"
+)
+
+type Service interface {
+	SetGroup(GoogleGroup)
+}
```
See above re: prone to human error

spiffxp commented on groups/service.go:
```diff
@@ -0,0 +1,415 @@
+}
+
+func (as *adminService) CreateOrUpdateGroupIfNescessary() error {
+	if *verbose {
+		log.Printf("createOrUpdateGroupIfNecessary %q", as.group.EmailId)
+	}
```
Logging style nit: I was trying to roughly follow the function signature before, and it's changed.  I'm not sure I like the way this suggestion looks, but moving without changing also means the log statement doesn't really line up with anything so makes me wonder what it should be instead
```suggestion
	if *verbose {
		log.Printf("adminService(%v).CreateOrUpdateGroupIfNecessary", as.group.EmailId)
	}
```

spiffxp commented on groups/service.go:
```diff
@@ -0,0 +1,415 @@
+	group, err := as.client.GetGroup(as.group.EmailId)
+	if err != nil {
+		if apierr, ok := err.(*googleapi.Error); ok && apierr.Code == http.StatusNotFound {
+			if !config.ConfirmChanges {
```
This is maybe a nice to have.  Keeping it like this is making it easier for me to diff what was moved, but I also think it'd be preferable to push config options like "dry run" and "verbose" into the service itself, so we're not tying back to globals.

spiffxp commented on groups/service.go:
```diff
@@ -0,0 +1,415 @@
+				}
+				g4, err := as.client.InsertGroup(&g)
+				if err != nil {
+					return fmt.Errorf("unable to add new group %q: %v", as.group.EmailId, err)
```
Out of scope for this PR, but for followup: it would be ideal if we used `%w` instead of `%v` for errs

spiffxp commented on groups/service.go:
```diff
@@ -0,0 +1,415 @@
+	return nil
+}
+
+func (as *adminService) CreateOrUpdateGroupIfNescessary() error {
```
Could be done here, could be done for followup: would like to see comments describing what these functions do

spiffxp commented on groups/service.go:
```diff
@@ -0,0 +1,415 @@
+	return nil
+}
+
+func (as *adminService) DeleteGroupsIfNecessary() error {
```
To keep harping on the oddity of the `SetGroup` thing... while other functions behave different depending on whether that was called, this does not.  The inconsistency feels off to me.

spiffxp commented on groups/service.go:
```diff
@@ -0,0 +1,415 @@
+	if err != nil {
+		return fmt.Errorf("unable to retrieve users in domain: %v", err)
+	}
+	if len(g.Groups) == 0 {
+		log.Println("No groups found.")
+		return nil
+	}
```
For followup since this is a straight refactor: we don't actually need this, since we'd skip by iterating over nothing below (and I don't think there's value in this logging statement)

spiffxp commented on groups/service.go:
```diff
@@ -0,0 +1,415 @@
+				break
+			}
+		}
+		if found || m.Role == "MEMBER" {
```
I know this is how you found it, but I scratched my head over this for a while.  Do you think you can add a comment explaining why this is needed? It's the only thing that makes `RemoveOwnersOrManagersFromGroup` different than `RemoveMembersFromGroup`

It would be nice to refactor this away as a followup.  I think this is to handle someone moving from manager to member, and not dropping them.  But making two passes over all group members seems like an inefficiency we'll want to remove

MadhavJivrajani replied:
> I think this is to handle someone moving from manager to member, and not dropping them.

If I understand correctly, this is along the lines of:

If a member m exists in our desired list of members, do nothing.
However, if this member m does not exist in our desired list of
members but is in the role of a MEMBER (non OWNER/MANAGER), still
do nothing since we care only about OWNER/MANAGER roles here.

spiffxp commented on groups/service.go:
```diff
@@ -0,0 +1,415 @@
+	as.group = group
+}
+
+func (as *adminService) AddOrUpdateGroupMembers(role string, members []string) error {
```
Same deal here re: returning `[]error` as a followup PR

spiffxp reviewed (changes requested):
It's been a while since I looked at all the code here holistically, so I called out a few things I would suggest as followups

Since this looks straightforward enough I'm OK if you get to the point of landing this before landing additional test coverage.  But it needs to run first ;)

```
go run reconcile.go
# command-line-arguments
./reconcile.go:211:15: undefined: AdminService
./reconcile.go:212:15: undefined: GroupService
./reconcile.go:216:13: undefined: NewAdminService
./reconcile.go:221:13: undefined: NewGroupService
./reconcile.go:350:19: cannot assign error to err in multiple assignment
```

MadhavJivrajani commented:
Oh wow, the compile time errors are odd... Will take a closer look soon, thanks for the reviews @spiffxp!

spiffxp commented:
I suspect we need to change the make target to build and run the binary

MadhavJivrajani commented:
@spiffxp I've addressed the review comments for the most part - the reviews that are still unresolved are to be made part of a follow-up PR.

Could you PTAL? :)

MadhavJivrajani commented:
I'll be adding the fake client and test coverage in subsequent PRs

spiffxp commented on groups/reconcile.go:
```diff
@@ -181,69 +178,172 @@ func main() {
+// It does so by making use of AdminService and GroupService which are mockable
+// interfaces.
+type Reconciler struct {
+	AdminSvc AdminService
+	GroupSvc GroupService
```
Can be for followup: I'm not sure these necessarily need to be public

spiffxp commented on groups/service.go:
```diff
@@ -0,0 +1,434 @@
+	DeleteGroupsIfNecessary() error
+	RemoveOwnerOrManagersFromGroup(group GoogleGroup, members []string) error
+	RemoveMembersFromGroup(group GoogleGroup, members []string) error
+	GetClient() AdminServiceClient
```
For followup: can we remove this? Not clear to me why we would want something other than the admin service (or the thing building it) to have access to the client

spiffxp commented on groups/service.go:
```diff
@@ -0,0 +1,434 @@
+// tasks using a GroupServiceClient.
+type GroupService interface {
+	UpdateGroupSettings(group GoogleGroup) error
+	GetClient() GroupServiceClient
```
For followup: can we remove this? Not clear to me why we would want something other than the group service (or the thing building it) to have access to the client

spiffxp commented on groups/reconcile.go:
```diff
@@ -181,69 +178,172 @@ func main() {
+			Name:        g.Name,
+			Description: g.Description,
+		}
+		g2, err := gsClient.Get(g.Email)
```
e.g. this could probably be a `GroupService` function so we don't need the client

spiffxp commented on groups/reconcile.go:
```diff
@@ -181,69 +178,172 @@ func main() {
+		group.Settings["WhoCanModerateMembers"] = g2.WhoCanModerateMembers
+		group.Settings["MembersCanPostAsTheGroup"] = g2.MembersCanPostAsTheGroup
+
+		l, err := asClient.ListMembers(g.Email)
```
ditto here, could be an `AdminService` function

spiffxp reviewed (approved):
`make run` is happy now

I've left the comments that can be done as followups unresolved

/approve
/lgtm

spiffxp commented on groups/reconcile.go:
```diff
@@ -181,69 +184,176 @@ func main() {
+	return &reconciler{adminService: as, groupService: gs}, nil
+}
+
+func (r *reconciler) reconcileGroups(groups []GoogleGroup) error {
```
Opened followup issue: https://github.com/kubernetes/k8s.io/issues/2772

MadhavJivrajani commented:
Thank you!
